Desktop Settings, VIAME CSV Parser, Pipeline Runner

subdavis opened the pull request:
Adds a settings page and starts to integrate with VIAME.

fixes #432

subdavis commented:
See #417 for screensots

subdavis commented:
This PR does not include a windows implementation or even platform switching logic.  However, since this thing is already huge, I'm not inclined to include those at this time.

BryonLewis commented on client/platform/desktop/backend/serializers/viame.ts:
```diff
@@ -0,0 +1,267 @@
+    /* Head */
+    const head = getCaptureGroups(HeadRegex, value);
+    if (head !== null) {
+      headTail[0] = [parseFloat(head[0]), parseFloat(head[1])];
+      geoFeatureCollection.features.push(_createGeoJsonFeature('Point', [headTail[0]], 'head'));
```
If you do the changes above for array[0] instead of array[1].splice(1) you need to modify these indices.  If not you would need to modify the attribute indicies.

BryonLewis commented on client/platform/desktop/constants.ts:
```diff
@@ -0,0 +1,88 @@
+  'image/gif',
+  'image/jpeg',
+  'image/png',
+  'image/svg+xml',
```
Are these really supported ?  Or are these a list of eventual things we want to support.  I don't believe the ImageAnnotator.js will support `svg`, or `bmp` types.  It should probably do gifs, but the others require a canvas and some additional logic we don't have right now.

BryonLewis reviewed (changes requested):
You got the core functionality there pretty quickly.  The web testing went really smooth and didn't notice any effects there (obviously you didn't really touch the web side).
 These are just minor things that I've noticed might need to be updated/changed.  Feel free to ask me any questions about my comments, I know that they aren't always clear.

subdavis commented on client/platform/desktop/backend/serializers/viame.ts:
```diff
@@ -0,0 +1,267 @@
+
+// eslint-disable-next-line import/no-extraneous-dependencies
+import csvparser from 'csv-parse';
+// eslint-disable-next-line import/no-extraneous-dependencies
```
The comment is all over the place, but I only want to do this in `platform/desktop` because I don't want any desktop-only dependencies to pollute the dependency list of `vue-media-annotator` and we are never going to expose desktop code as an NPM library, so this abuse won't affect anyone but us.

I'd be OK with that rule as long as it goes in `platform/desktop/.eslintrc` instead of package.json

subdavis commented on client/package.json:
```diff
@@ -7,16 +7,22 @@
     "@types/jest": "^25.2.3"
   },
   "files": [
+    "/bin/",
```
[no](https://stackoverflow.com/questions/24942161/does-npm-ignore-files-listed-in-gitignore)

subdavis commented:
@BryonLewis you were right about pretty much everything except gitignore: that would exclude the asset from getting published.

* Addressed comments in c64252
* Tested polygon parsing, head tail, etc. with your suggestions and it looks like it works.
* More robust test suite is needed.

BryonLewis commented on client/platform/desktop/backend/serializers/viame.ts:
```diff
@@ -0,0 +1,273 @@
+    /* Polygon */
+    const poly = getCaptureGroups(PolyRegex, value);
+    if (poly !== null) {
+      const coords: number[][] = [];
+      const polyList = poly[2].split(' ');
+      polyList.forEach((coord, j) => {
+        if (j % 2 === 0) {
+          // Filter out ODDs
+          if (polyList[j + 1]) {
+            coords.push([parseFloat(coord), parseFloat(polyList[j + 1])]);
+          }
+        }
+      });
+      geoFeatureCollection.features.push(_createGeoJsonFeature('Polygon', coords));
+    }
+  });
```
Just a reminder to eventually add in some better checking here for `poly` being an array with an even number of coordinates.  (I think that should be a requirement)

subdavis replied:
Captured in https://github.com/VIAME/VIAME-Web/issues/444

BryonLewis commented on client/platform/desktop/backend/serializers/viame.ts:
```diff
@@ -0,0 +1,273 @@
+  if (value === 'false') {
+    return false;
+  }
+  try {
+    return parseFloat(value);
+  } catch {
+    return value;
+  }
+}
```
This may need to be modified because It will return `NaN` for some string values if it can't properly parse the float.

subdavis replied:
I forgot how the parseFloat api worked.  It doesn't throw an error, it just returns NaN.

BryonLewis reviewed (changes requested):
All the fixes are tested and work well.
There is just a minor thing related to using Strings as Detection/Track Attributes.  I believe right now certain values will return NaN.

BryonLewis reviewed (approved):
Pulled/Tested and now string based attributes show up properly.
